client/apps/edit/components/content/sections/text/index.jsx

```diff
@@ -0,0 +1,557 @@
+import { TextInputUrl } from 'client/components/rich_text/components/input_url'
+import { TextNav } from 'client/components/rich_text/components/text_nav'
+import * as Config from './draft_config'
+
+export class SectionText extends Component {
+  static propTypes = {
+    article: PropTypes.object,
+    editing: PropTypes.bool,
+    hasFeatures: PropTypes.bool,
+    index: PropTypes.number,
+    onChange: PropTypes.func,
+    isContentEnd: PropTypes.bool,
+    isContentStart: PropTypes.bool,
+    onSetEditing: PropTypes.func,
+    section: PropTypes.object,
+    sections: PropTypes.object
+  }
+
+  constructor (props) {
+    super(props)
+
+    this.state = {
+      editorState: EditorState.createEmpty(
+        Config.composedDecorator(props.article.layout)
+      ),
+      focus: false,
+      html: null,
+      selectionTarget: null,
+      showMenu: false,
+      showUrlInput: false,
+      plugin: null,
+      url: null
+    }
+
+    props.sections.on('change:autolink', this.editorStateFromProps)
+  }
+
+  componentWillMount = () => {
+    const { editing, section } = this.props
+
+    if (section.body && section.body.length) {
+      this.setEditorStateFromProps()
+    } else if (editing) {
+      this.focus()
+    }
+  }
+
+  componentDidUpdate = (prevProps) => {
+    const {
+      article,
+      editing,
+      isContentEnd,
+      onChange,
+      section
+    } = this.props
+
+    // Reset contentEnd markers if end has changed
+    if (isContentEnd !== prevProps.isContentEnd) {
+      if (article.layout !== 'classic') {
+        const html = setContentEnd(section.body, isContentEnd)
+        onChange('body', html)
+      }
+    }
+    // Focus/blur editor if editing prop has changed
+    // For change of section via key commands (handleTab, splitSection)
+    if (editing && editing !== prevProps.editing) {
+      this.focus()
+    } else if (!editing && editing !== prevProps.editing) {
+      if (this.domEditor) {
+        this.domEditor.blur()
+      }
+    }
+  }
+
+  // EDITOR AND CONTENT STATE
+  setEditorStateFromProps = () => {
+    const {
+      article,
+      editing,
+      isContentEnd,
+      section
+    } = this.props
+
+    let html = standardizeSpacing(section.body)
+
+    if (article.layout !== 'classic') {
+      html = setContentEnd(html, isContentEnd)
+    }
+
+    const blocksFromHTML = convertFromRichHtml(html)
+    const decorators = Config.composedDecorator(article.layout)
+    let editorState = EditorState.createWithContent(blocksFromHTML, decorators)
+
+    if (editing) {
+      editorState = setSelectionToStart(editorState)
+    }
+
+    this.setState({ editorState, html })
+  }
+
+  onChange = (editorState) => {
+    const { article, onChange, section } = this.props
+    const html = convertToRichHtml(editorState, article.layout)
+
+    if (section.body !== html) {
+      onChange('body', html)
+    }
+    this.setState({ editorState, html })
```

Comment:
If we move this up to the if block, I think we could prevent an additional re-render?

Reply:
EditorState keeps track of changes in focus and text selection, so we still need to setState -- the if block is to prevent unnecessary saves when those events occur.